File: CSES/treeDiameter.py
from collections import deque
def bfs(ady, s):
	vis = set()
	dist = {}
	q = deque([])
	q.append(s); vis.add(s); dist[s]=0

	while q:
		v = q.popleft()
		for u in ady[v]:
			if u not in vis:
				vis.add(u)
				q.append(u)
				dist[u] = dist[v]+1
	return dist


n = int(input())
t = {i:set() for i in range(n)}
for _ in range(n-1):
	a,b = map(int,input().split())
	t[a-1].add(b-1); t[b-1].add(a-1)

dist = bfs(t,0)
new_src = 0
M = 0
for i in dist:
	if(dist[i]>M):
		M=dist[i]; new_src=i
dist = bfs(t,new_src)
ans = 0
for i in dist: ans=max(ans,dist[i])
print(ans)

File: CSES/Tree/treeDiameter.cpp
#include <bits/stdc++.h>
using namespace std;
#define pb push_back
#define sz size()
#define forn(i,n) for(int i=0; i<n; i++)

void bfs(int n, vector<int> ady[], int dist[], int s){
	queue<int> q;
	bool vis[n]={false};
	q.push(s); vis[s]=true; dist[s]=0;

	while(!q.empty()){
		int v = q.front();
		q.pop();
		for(auto u: ady[v]){
			if(!vis[u]){
				vis[u]=true;
				q.push(u);
				dist[u]=dist[v]+1;
			}
		}
	}
}

int main(){

	ios::sync_with_stdio(0);
	cin.tie(0);
	int n; cin >> n;
	vector<int> t[n];
	forn(i,n-1){
		int a,b; cin >> a >> b; a--;b--;
		t[a].pb(b);t[b].pb(a);
	}
	int dist1[n]={0};
	bfs(n,t,dist1,0);

	int M=0,Midx=0;
	forn(i,n){
		if(dist1[i]>M){
			M = dist1[i]; Midx=i;
		}
	}
	int dist2[n]={0};
	bfs(n,t,dist2,Midx);
	int ans = 0;
	forn(i,n) ans = max(ans,dist2[i]);
	cout << ans << endl;
	return 0;
}
